### DISCUSSION TITLE
Is it possible to use the Pomelo EF provider with Quart.NET

### DISCUSSION THREAD

**Discussion by jussityynela** [quartznet/quartznet#1362] [Q&A]
It seems that the connector that the Quart.NET supports out of the box is the MySQL.Data ADO.net driver
We are using the Pomelo EF library which uses the MySqlConnector ADO.Net provider, hence it
is incompatible with the MySQL.Data ( when trying to use the MySQL with Quart.NET it fails because it cannot find MySQL.Data assembly)

Can use some generic provider approach to tell Quart.NET to use MySQLConnector or would it be easier to make a 
fork to achieve it? If I need to go to the fork approach, is there any gotcha's that I should be aware of?

I just realised that It _might_ be possible to use the MySQL.Data and MySqlConnector side by side, but don't know what would be the caveats of doing so.

**Comment by lahma** [comment]
You need to change provider to `MySqlConnector`. You can find supported provider names here: https://github.com/quartznet/quartznet/blob/main/src/Quartz/Impl/AdoJobStore/Common/dbproviders.netstandard.properties

The code config (SchedulerBuilder, MS DI) default to using MySql.Data so you need to set the provider property.

**Reply by lahma** [reply]
Related discussion: https://github.com/quartznet/quartznet/discussions/1342

**Reply by jussityynela** [reply]
Blast,  I checked that file, and somehow glanced over that provider. I apologize for the inconvenience.
Anyway this is how I set it up now. 
```C#
             var connectionString = Environment.GetEnvironmentVariable("connectionString") ?? Configuration["ConnectionStrings:Training"];
            services.AddQuartz(q =>
                {
                    //....
                    q.UsePersistentStore(x =>
                       {
                           x.UseGenericDatabase("MySqlConnector", db => db.ConnectionString = connectionString);
                           x.UseJsonSerializer();
                       }
                    );
                  //...
                }

``` 
Did I understood correctly, Is it possible to use .UserMySQL and then set the provider

**Reply by lahma** [reply]
I think this should work:

```csharp
q.UsePersistentStore(s =>
{
    s.UseProperties = true;
    s.UseMySql(options =>
    {
        options.ConnectionString = "some connection string";
    });
    s.SetProperty($"quartz.dataSource.{SchedulerBuilder.AdoProviderOptions.DefaultDataSourceName}.provider", "MySqlConnector");
    s.UseJsonSerializer();
})
```
The fluent API is wrapper over the properties so you can override already set ones on next step. So your version is pretty much same except that `UseMySql` also configures correct driver delegate which you are now missing.

A PR is always welcome to to add `Provider` property to `AdoProviderOptions` as MySql can use two different ones 😉

**Reply by fsaffioti** [reply]
@lahma but with this configuration, I can reuse the same connection pool that I'm using for my app, I mean, my API controllers ?
I have API controllers that receive data and immediately schedule a new job, and i'm using MySQL database (with mysqlconnector)

I was using 2 different string connection with the same name of the database but with different aplicationName
quartz db string configuration :
 "quartz.dataSource.myDS.connectionString": "{ConnectionDB:BaseConnectionString};database={ConnectionDB:Database};ApplicationName={Quartz:quartz.scheduler.instanceName};" 
Controllers db connection:
    "MainConnectionString": "{ConnectionDB:BaseConnectionString};database={ConnectionDB:Database};ApplicationName ={AppConfig:ServerName};"


I noticed that every time I receive data on my API controller, I open a DB connection, but also Quartz open a  DB connection
so with 30 parallel requests, I have 30 connections opened from my Api Controller and 30 connections opened from Quartz, but then  the background  process triggered by Quartz are a couple of them every second, so I have a lot of Quartz connections no more used
so I would reuse these connections opened by Quartz on other API controllers.

**Reply by lahma** [reply]
Is this a theoretical issue or a real issue? Like is 60 connections too much for your db server?

Pooling should ensure that open connections are reused so they are left open waiting to be acquired.

You could use the same connection string if you want to. It seems that you are using same database for your business app and Quartz so load probably isn't that high. Dedicated database for quartz ensures less resource contention and interference from backups etc.

I think this is more like a MySqlConnector configuration issue. There seems to  be a wealth of options to set: https://mysqlconnector.net/connection-options/

**Reply by fsaffioti** [reply]
Yes it's a real issue, now I'm testing with rds mysql micro instance that allow max 62 connections, but if i cannot reuse connection pool opened by quartz and no more used i have to divide by 2 the max connection available, maybe using the same connection string also with the application name mysqlconnector will be able to reuse connection pool between business app and quartz.
If i use different databases  a connection to one database will never be able to be re-used for another database.

**Reply by lahma** [reply]
Then configure the max pool size in the connection string(s), disabling pooling might be a bit too drastic.

**Reply by fsaffioti** [reply]
In stack overflow i saw also DBConnectionManager.getInstance().getConnection("myDS") in c# src/Quartz/Util/DBConnectionManager.cs
In this case i think i can use the same data source and pool connection, make sense?
But first i will try the same connection string  so i don't have to change mysql api

**Reply by lahma** [reply]
Trying to pool outside of connection library is probably a bad idea, there's so many things to watch out for.
